pilot/pkg/model/service.go

```diff
@@ -1286,7 +1276,11 @@ func (s *Service) GetAddressForProxy(node *Proxy) string {
 		}
 	}
 
-	return s.DefaultAddress
+	addresses := filterAddresses([]string{s.DefaultAddress}, node.SupportsIPv4(), node.SupportsIPv6())
```

Comment:
I am a bit concerned by this.

This goes back many years, not new code with DUAL_STACK or ambient or anything.

Before, you could make a SE like 

```yaml
apiVersion: networking.istio.io/v1
kind: ServiceEntry
metadata:
  name: example
spec:
  addresses: ["::1"]
  hosts:
  - httpbin.org
  location: MESH_EXTERNAL
  ports:
  - number: 1234
    name: TCP
    protocol: TCP
  resolution: DNS
```

and it would apply to all sidecars. 

Is it likely to ever be hit... probably not, but could be. A node may actually support IPv6 but not report it as one of its IPs (unlikely on K8s, but maybe on VMs) -- though that is a weird case I am not too concerned about.

My bigger concern is that before, this SE is just basically a NOP; the client can never reach it (they would get `Network is unreachable` and/or not have ip6tables setup). Now it is under ANY IPv4 address. This can blackhole all 'port 1234' traffic for instance.

Another behavioral change here is in the DNS proxy. Before, with this config we would return an AAAA query. Without it we do not. Given we are _probably_ in a scenario where they cannot reach it anyways, they _probably_ don't care to do a DNS request to it... but they could, and we are changing the behavior there in ways I think are unexpected.

Reply:
Just to be clear, I think its weird we return an unsupported IP family -- possibly even a bug. But this is also something we have done since the beginning of Istio, and I worry fixing it will introduce new issues.

If we do decide to filter them out, we must do so in a different way than returning `UnspecifiedIP` which is unsafe

Reply:
Agreed,  this is trying to fix a very rare case. Say we have two clusters, one is dual stack, the other is ipv6.

And a service (dual stack ) only created in cluster1, so a proxy in cluster two will get an ipv4 address.  So IMO we may need to document, we donot support multi clusters with different setup. Even for vms, they must support same ipfamilies with cluster.

Reply:
I don't think it's really a problem to get an ipv4 IP on an Ipv6 cluster tbh... it will just not match any traffic but doesn't cause any harm?

Reply:
I am afraid the bind may fail. And actually if we return `0.0.0.0` here, in the listener builder, there is some logic determin the bind address by proxy ip mode. It would choose :: if proxy support ipv6

Reply:
for the bind it matters but I am not sure the binding ever hits this code?

Reply:
In the grpcGen

Reply:
gRPC doesn't bind to anything in any scenario

Reply:
For the sidecar, also same here

```
	switch listenerProtocol {
	case istionetworking.ListenerProtocolTCP, istionetworking.ListenerProtocolAuto:
		// Determine the listener address if bind is empty
		// we listen on the service VIP if and only
		// if the address is an IP address. If its a CIDR, we listen on
		// 0.0.0.0, and setup a filter chain match for the CIDR range.
		// As a small optimization, CIDRs with /32 prefix will be converted
		// into listener address so that there is a dedicated listener for this
		// ip:port. This will reduce the impact of a listener reload
		if listenerOpts.bind.Primary() == "" { // TODO: make this better
			svcListenAddress := listenerOpts.service.GetAddressForProxy(listenerOpts.proxy)
			svcExtraListenAddresses := listenerOpts.service.GetExtraAddressesForProxy(listenerOpts.proxy)
			// Override the svcListenAddress, using the proxy ipFamily, for cases where the ipFamily cannot be detected easily.
			// For example: due to the possibility of using hostnames instead of ips in ServiceEntry,
			// it is hard to detect ipFamily for such services.
			if listenerOpts.service.Attributes.ServiceRegistry == provider.External && listenerOpts.proxy.IsIPv6() &&
				svcListenAddress == constants.UnspecifiedIP {
				svcListenAddress = constants.UnspecifiedIPv6
			}

			// For dualstack proxies we need to add the unspecifed ipv6 address to the list of extra listen addresses
			if listenerOpts.service.Attributes.ServiceRegistry == provider.External && listenerOpts.proxy.IsDualStack() &&
				svcListenAddress == constants.UnspecifiedIP {
				svcExtraListenAddresses = append(svcExtraListenAddresses, constants.UnspecifiedIPv6)
			}
			// We should never get an empty address.
			// This is a safety guard, in case some platform adapter isn't doing things
			// properly
			if len(svcListenAddress) > 0 {
				if !strings.Contains(svcListenAddress, "/") {
					listenerOpts.bind.binds = append([]string{svcListenAddress}, svcExtraListenAddresses...)
				} else {
					// Address is a CIDR. Fall back to 0.0.0.0 and
					// filter chain match
					listenerOpts.bind.binds = actualWildcards
					listenerOpts.cidr = append([]string{svcListenAddress}, svcExtraListenAddresses...)
				}
			}
		}
```

Reply:
Are those actually binding though? I thought they just end up as bind_to_port=false listeners?

Reply:
maybe

Reply:
The branches numbers are really large, i cannot lookinto all of them. So i agree keep as before may be more safe